classes/boardmanager.php

```diff
@@ -0,0 +1,1041 @@
+
+            $this->formatter->put('board', $update);
+            $this->formatter->put('columns', $data);
+            return $data['id'];
+        }
+        return 0;
+    }
+
+
+    /**
+     * Adds a new card.
+     *
+     * @param int $columnid Id of the column
+     * @param int $aftercard Id of the card before (0 means to insert at top)
+     * @param array $data Data to override default values
+     * @return int Id of the card
+     */
+    public function add_card(int $columnid, int $aftercard = 0, array $data = []): int {
+        global $DB, $USER;
+        $aftercard = intval($aftercard);
+        $defaults = [
+            'title' => get_string('newcard', 'mod_kanban'),
+            'options' => '{}',
+            'description' => '',
+            'createdby' => $USER->id
+        ];
+        $defaultsfixed = [
+            'kanban_board' => $this->board->id,
+            'kanban_column' => $columnid,
+            'timecreated' => time(),
+            'timemodified' => time(),
+            'sequence' => '',
+        ];
+        $data = array_merge($defaults, $data, $defaultsfixed);
+
+        $data['id'] = $DB->insert_record('kanban_card', $data);
+        $data['assignees'] = [];
+
+        $column = $DB->get_record('kanban_column', ['id' => $columnid]);
+
+        $update = [
+            'id' => $columnid,
+            'sequence' => helper::sequence_add_after($column->sequence, $aftercard, $data['id']),
+            'timemodified' => time()
+        ];
+        $DB->update_record('kanban_column', $update);
+
+        // Users can always edit cards they created.
+        $data['canedit'] = true;
+        $data['columnname'] = $column->title;
+
+        $this->formatter->put('cards', $data);
+        $this->formatter->put('columns', $update);
+        $this->write_history('added', constants::MOD_KANBAN_CARD, $data, $columnid, $data['id']);
+        helper::update_cached_timestamp($this->board->id, constants::MOD_KANBAN_COLUMN, $update['timemodified']);
+        helper::update_cached_timestamp($this->board->id, constants::MOD_KANBAN_CARD, $update['timemodified']);
+
+        return $data['id'];
+    }
+
+    /**
+     * Moves a column.
+     *
+     * @param int $columnid Id of the column to move
+     * @param int $aftercol Id of the (future) column before (0 means to move at the leftmost position)
+     * @return void
+     */
+    public function move_column(int $columnid, int $aftercol) {
+        global $DB;
+        $column = $DB->get_record('kanban_column', ['id' => $columnid]);
+        if (!$this->board->locked && !$column->locked) {
+            $update = [
+                'id' => $this->board->id,
+                'sequence' => helper::sequence_move_after($this->board->sequence, $aftercol, $columnid),
+                'timemodified' => time(),
+            ];
+            $DB->update_record('kanban_board', $update);
+            helper::update_cached_board($update['id']);
```

Comment:
isn't it also necessary to also update the cached timestamp here?

Reply:
No, timestamp is only cached for columns / cards. Moved column is not updated here.